File: KFramework.WebGame/KFramework.MonoGame/HTML/HTML_Canvas.cs
using System;

namespace KFramework.MonoGame
{
    /// <summary>
    /// 一块 HTML 画布（<c>&lt;canvas&gt;</c>）—— 与 TS 层 <c>html_canvas.ts</c> 的 <c>canvas</c> 模块一一对应的对象封装。
    /// 一个实例 = 页面上的一块画布（按 DOM id 定位），自带当前矩形 <see cref="Rect"/>，调用方不必每次重复传 id。底层直接转发到 <see cref="JSBind_HTML_Canvas"/>。
    /// <para>
    /// 所有写操作只改 CSS：后备缓冲 = CSS 尺寸 × DPR，由 <see cref="GraphicsDevice.SyncCanvasSize"/> 同步，
    /// 输入坐标也按画布 rect 自动换算。
    /// </para>
    /// </summary>
    /// <remarks>
    /// 与 TS 导出的对应关系：create / applyLayout / restoreLayout / destroy / exists / getRect / getHTMLPageSize，
    /// 全部经 <see cref="JSBind_HTML_Canvas"/> 的 [JSImport] 直连，本类不再经 <see cref="HTML_Canvas_Func"/> 中转
    /// （<see cref="HTML_Canvas_Func"/> 现在只保留「建」与「销」两块画布的生命周期入口）。
    /// </remarks>
    public sealed class HTML_Canvas
    {
        private Rectangle? _rect;

        /// <summary>
        /// 用 DOM id（或 "#id" 选择器）关联一块画布。构造本身不创建 DOM 元素，
        /// 需要新建时调用 <see cref="Create(HTML_CanvasLayoutMode, int, int, int, int)"/>。
        /// </summary>
        /// <param name="idOrSelector">画布 DOM id，可写 "#id" 形式；空串回落到默认 id。</param>
        public HTML_Canvas(string idOrSelector)
        {
            Id = HTML_Canvas_Func.ToCanvasId(idOrSelector);
        }

        /// <summary>本实例对应的 DOM id（已归一化，无 "#" 前缀）。</summary>
        public string Id { get; }

        /// <summary>页面上这块画布是否已存在（含引擎曾创建过的）。</summary>
        public bool Exists => JSBind_HTML_Canvas.Exists(Id);

        /// <summary>
        /// 画布当前的实际矩形（相对窗口左上角，单位 CSS 像素）—— 对应 HTML 里这块元素的 rect。
        /// </summary>
        /// <remarks>
        /// 值是从浏览器读回来的缓存：每次布局操作成功后会自动刷新；
        /// 若画布被页面自己的 CSS / 脚本改动过，调 <see cref="Refresh"/> 重新拉取。
        /// 画布不存在时为 <see cref="Rectangle.Empty"/>。
        /// </remarks>
        public Rectangle Rect => _rect ??= ReadRect() ?? Rectangle.Empty;

        /// <summary>重新从浏览器读回 <see cref="Rect"/>（画布被外部改动后调用）。</summary>
        public void Refresh() => _rect = ReadRect();

        /// <summary>创建这块画布：按 <paramref name="mode"/> 布局（Rect 摆位 / Centered 居中 / Fullscreen 填满整个 HTML 页面 / Size 仅设尺寸）。已存在时返回 false。</summary>
        public bool Create(HTML_CanvasLayoutMode mode, int x, int y, int width, int height)
            => Refresh(JSBind_HTML_Canvas.Create(Id, (int)mode, x, y, width, height));

        /// <summary>
        /// 【通用布局入口】一次调用表达四种摆位需求，对应 TS 的 applyLayout。
        /// 引擎内部与上层业务都可以通过它组合出任意布局，无需为每种情况新增一个 JS 导出。
        /// </summary>
        /// <param name="mode">布局方式。</param>
        /// <param name="x">左上角 X（<see cref="HTML_CanvasLayoutMode.Centered"/> / <see cref="HTML_CanvasLayoutMode.Size"/> 时忽略）。</param>
        /// <param name="y">左上角 Y（同上）。</param>
        /// <param name="width">CSS 宽度（<see cref="HTML_CanvasLayoutMode.Fullscreen"/> 时忽略）。</param>
        /// <param name="height">CSS 高度（同上）。</param>
        /// <returns>画布存在且设置成功时返回 true。</returns>
        public bool SetLayout(HTML_CanvasLayoutMode mode, int x, int y, int width, int height)
            => Refresh(JSBind_HTML_Canvas.ApplyLayout(Id, (int)mode, x, y, width, height));

        /// <summary>设置画布位置与 CSS 尺寸（像素，相对窗口左上角）。</summary>
        public bool SetRect(int x, int y, int width, int height)
            => SetLayout(HTML_CanvasLayoutMode.Rect, x, y, width, height);

        /// <summary>只改 CSS 尺寸，位置保持不动。</summary>
        public bool SetSize(int width, int height)
            => SetLayout(HTML_CanvasLayoutMode.Size, 0, 0, width, height);

        /// <summary>摆到浏览器窗口正中；之后窗口缩放会自动重新居中。</summary>
        public bool SetCentered(int width, int height)
            => SetLayout(HTML_CanvasLayoutMode.Centered, 0, 0, width, height);

        /// <summary>撤销引擎写在这块画布上的行内样式，恢复页面自身布局，并解除居中跟随。</summary>
        public bool RestoreLayout() => Refresh(JSBind_HTML_Canvas.RestoreLayout(Id));

        /// <summary>从 DOM 移除这块画布。</summary>
        public bool Destroy()
        {
            bool ok = JSBind_HTML_Canvas.Destroy(Id);
            if (ok) _rect = null;
            return ok;
        }

        /// <summary>写操作成功后顺带回读一次真实矩形，保证 <see cref="Rect"/> 与浏览器一致。</summary>
        private bool Refresh(bool ok)
        {
            if (ok) Refresh();
            return ok;
        }

        /// <summary>从浏览器读回画布矩形；画布不存在或尺寸非法时返回 <c>null</c>。</summary>
        private Rectangle? ReadRect()
        {
            Span<int> view = stackalloc int[4];
            JSBind_HTML_Canvas.GetRect(Id, view);

            Rectangle rect = new(view[0], view[1], view[2], view[3]);
            return rect.Width <= 0 || rect.Height <= 0 ? null : rect;
        }

        /// <inheritdoc />
        public override string ToString() => $"HTML_Canvas(id={Id}, rect={Rect})";
    }
}

File: KFramework.Mir/WebGame.Mir2/WebGame.Mir2.MonoGame.Client/wwwroot/jsengine/html_canvas.js
// 【依赖 C#】由 KFramework.MonoGame.JSBind_HTML_Canvas 经 [JSImport(module: "canvas")] 调用；产物 html_canvas.js 由 SyncJsEngine 复制。
// 画布元素管理层（HTML5 <canvas>）：创建、设置位置与尺寸、删除，以及「页面里没有画布时自动建一块全屏默认画布」。
//
// 与 gl.ts 的分工：
//   · 本文件只管 **DOM 元素**（canvas 有没有、插在哪、CSS 多大、删没删）；
//   · gl.ts 只管 **WebGL2 上下文**（从本文件给出的元素上 getContext('webgl2')）。
// 因此 gl.ts 不再自己 querySelector，避免出现两套查找 / 创建规则。
//
// 尺寸链路（改动这里之前请先读 platform.js 的 getCanvasSize 注释）：
//   CSS 尺寸（本文件写的 style.width / style.height）
//     → platform.getCanvasSize 每帧读 rect × DPR 写进 canvas.width / canvas.height（backing）
//     → C# 侧 GraphicsDevice.SyncCanvasSize 更新 PresentationParameters 与 Viewport，并触发 GameWindow.SizeChanged。
// 所以本文件只改 CSS，backing / 视口 / 输入坐标（input_common 按 rect 换算）会在下一帧自动跟上。
/** id → 画布元素。由本模块统一持有，删除时同步摘除。 */
const canvases = new Map();
/** 未指定画布时的默认 DOM id（与 C# 侧 Game 的默认选择器 "#game" 对齐）。 */
export const DEFAULT_CANVAS_ID = 'game';
/**
 * 把 "#game" 这类选择器归一化成 DOM id；空串回落到默认 id。
 * C# 侧传进来的永远是选择器或 id 字符串，这里做一次统一处理。
 */
function toId(idOrSelector) {
    const trimmed = (idOrSelector ?? '').trim().replace(/^#/, '');
    return trimmed.length > 0 ? trimmed : DEFAULT_CANVAS_ID;
}
/** 所有画布共用的样式（全屏与矩形定位只有 left / top / width / height 不同）。 */
function applyCommonStyle(element) {
    element.style.position = 'fixed';
    element.style.display = 'block';
    element.style.margin = '0';
    element.style.padding = '0';
    element.style.outline = 'none';
    element.style.touchAction = 'none';
    element.style.zIndex = '0';
}
/** 铺满视口（100% / inset 效果由 left/top=0 + width/height=100% 实现）。 */
function applyFullscreenStyle(element) {
    applyCommonStyle(element);
    element.style.left = '0px';
    element.style.top = '0px';
    element.style.width = '100%';
    element.style.height = '100%';
}
/** 绝对定位到 (x, y)，CSS 尺寸为 width × height（单位：CSS 像素，最小 1px）。 */
function applyRectStyle(element, x, y, width, height) {
    applyCommonStyle(element);
    element.style.left = `${Math.round(x)}px`;
    element.style.top = `${Math.round(y)}px`;
    element.style.width = `${Math.max(1, Math.round(width))}px`;
    element.style.height = `${Math.max(1, Math.round(height))}px`;
}
function lookup(id) {
    const known = canvases.get(id);
    if (known)
        return known;
    const byDom = document.getElementById(id);
    if (byDom instanceof HTMLCanvasElement) {
        canvases.set(id, byDom);
        return byDom;
    }
    return null;
}
/**
 * 创建一块全屏画布并插入 <body>。
 * @param idOrSelector 画布 id（可写 "#id" 形式）；已存在同名画布时返回 false。
 */
export function createFullscreen(idOrSelector) {
    const id = toId(idOrSelector);
    if (lookup(id))
        return false;
    const element = document.createElement('canvas');
    element.id = id;
    applyFullscreenStyle(element);
    document.body.appendChild(element);
    canvases.set(id, element);
    return true;
}
/**
 * 在指定位置创建一块画布并插入 <body>。
 * @param idOrSelector 画布 id（可写 "#id" 形式）；已存在同名画布时返回 false。
 * @param x 左上角 X（CSS 像素，相对视口）。
 * @param y 左上角 Y（CSS 像素，相对视口）。
 * @param width CSS 宽度（像素）。
 * @param height CSS 高度（像素）。
 */
export function create(idOrSelector, x, y, width, height) {
    const id = toId(idOrSelector);
    if (lookup(id))
        return false;
    const element = document.createElement('canvas');
    element.id = id;
    applyRectStyle(element, x, y, width, height);
    document.body.appendChild(element);
    canvases.set(id, element);
    return true;
}
/** 设置已有画布的位置与 CSS 尺寸。没有这块画布时返回 false。 */
export function setRect(idOrSelector, x, y, width, height) {
    const element = lookup(toId(idOrSelector));
    if (!element)
        return false;
    applyRectStyle(element, x, y, width, height);
    return true;
}
/** 把已有画布恢复为铺满视口。没有这块画布时返回 false。 */
export function setFullscreen(idOrSelector) {
    const element = lookup(toId(idOrSelector));
    if (!element)
        return false;
    applyFullscreenStyle(element);
    return true;
}
/** 删除画布：从 DOM 移除并注销。没有这块画布时返回 false。 */
export function destroy(idOrSelector) {
    const id = toId(idOrSelector);
    const element = lookup(id);
    if (!element)
        return false;
    element.remove();
    canvases.delete(id);
    return true;
}
/** 该画布是否存在（页面上已有、或由本模块创建过）。 */
export function exists(idOrSelector) {
    return lookup(toId(idOrSelector)) !== null;
}
/** 取画布元素本身（给 gl.ts 初始化上下文用，不由 C# 直接调用）。 */
export function getCanvas(idOrSelector) {
    return lookup(toId(idOrSelector));
}
/**
 * 解析「游戏要用哪块画布」：页面里已有同 id 的元素就直接用；
 * 没有则由引擎自建一块全屏默认画布——这就是「游戏启动时自动创建默认全屏画布」的入口。
 * @param idOrSelector 画布 id 或选择器（空串 / 缺省 → 默认 id）。
 * @returns 画布元素；无法创建时为 null。
 */
export function resolveCanvasElement(idOrSelector) {
    const id = toId(idOrSelector);
    const existing = lookup(id);
    if (existing)
        return existing;
    if (!createFullscreen(id))
        return null;
    return canvases.get(id) ?? null;
}
